libcobblersignatures/models/osversion.py

```diff
@@ -6,20 +6,47 @@ class OsArchitectures(Enum):
     An enumeration which defines the in Cobbler available architectures.
     """
     i386 = 1
+    """
+    32-bit architecture which is also called ``IA-32`` or ``80x86`` by some people.
+    """
     x86_64 = 2
+    """
+    64-bit architecture which is also called ``x64``, ``x86-64``, ``AMD64`` or ``amd64``.
+    """
     ppc = 3
+    """
+    32-bit big-endian PowerPC architecture.
+    """
     ppc64 = 4
+    """
+    64-bit big-endian PowerPC architecture.
+    """
     amd64 = 5
+    """
+    Synonym for ``x86_64``.
+    """
 
 
 class RepositoryBreeds(Enum):
     """
     An enumeration which defines the in Cobbler available repository breeds.
     """
     rsync = 1
+    """
+    A repository which is synced by rsync.
+    """
     rhn = 2
+    """
+    A repository type from Red Hat which can be used by yum.
+    """
     yum = 3
+    """
+    A repository which is manged by yum.
+    """
     apt = 4
+    """
+    A repository which is managed by apt.
+    """
 
 
 class Osversion:
```

Comment:
If you ask me, this class really wants to be a [`dataclass`](https://docs.python.org/3/library/dataclasses.html#dataclasses.dataclass). I wonder how much mutability you need, are the deleters and setters really needed? 

I could imagine something like the below is working as well, but I don't have the full context of theses models:
```python
@dataclasses.dataclass
class OSVersion:  # pylint: disable = too-many-instance-attributes
    """Version of an OS breed."""

    signatures: list
    version_file: str
    version_file_regex: str
    kernel_arch: str
    kernel_arch_regex: str
    supported_arches: list
    supported_repo_breeds: list
    kernel_file: str
    initrd_file: str
    isolinux_ok: bool
    default_autoinstall: str
    kernel_options: str
    kernel_options_post: str
    template_files: str
    boot_files: list
    boot_loaders: dict

    def encode(self) -> dict:
        """
        Encodes a single :class:`Osversion`. This means that the properties of an object is transferred into a JSON.

        :return: The dictionary with the data.
        """
        return {
            "signatures": self.signatures,
            "version_file": self.version_file,
            "version_file_regex": self.version_file_regex,
            "kernel_arch": self.kernel_arch,
            "kernel_arch_regex": self.kernel_arch_regex,
            "supported_arches": self.supported_arches,
            "supported_repo_breeds": self.supported_repo_breeds,
            "kernel_file": self.kernel_file,
            "initrd_file": self.initrd_file,
            "isolinux_ok": self.isolinux_ok,
            "default_autoinstall": self.default_autoinstall,
            "kernel_options": self.kernel_options,
            "kernel_options_post": self.kernel_options_post,
            "template_files": self.template_files,
            "boot_files": self.boot_files,
            "boot_loaders": self.boot_loaders,
        }

# instead of
version = Osversion()
version.decode(my_dict)
# you'd use
version = OSVersion(**my_dict)
```

The upside of using dataclasses is that you only need to write what attributes you have and what type they should have and the rest is autogenerated. But you'd lose strict type checking at runtime, at least without additional code/libraries.

Reply:
Well what is missing currently is the fact that for example I want to do more checks in the setters. For example I do want to do regex validation and give back the actual objects on the getters later. Also I don't want consumers to be able to acutally delete attributes, thus the explicit overwrite of the deleters. Also this is a nice solution in my eyes to have a possibility to reset values to their defaults if required.

Tbh. I didn't knew the concept of a dataclass before but I think especially for the dict and list the validation should be there at least which then means that a dataclass is not enough. What I like is that the objects can be made immutable but since I am using lists and dicts that sadly is not possible to my understanding.

If I only would limit myself to type checking I found this library - https://pypi.org/project/dataclass-type-validator/ - but I am not sure if it is wise to do only minimal validation if you have user input... My design skills are sadly not my biggest strength.

Reply:
Summary of a talk @agraul @meaksh and I had after our standup: Since the only consumers of this library are the CLI and Cobbler itself, currently and in the foreseeable future, it makes more sense to handle the data validation directly in these points. This point is strengthened by the fact that it doesn't matter how much validation I will build in because in Python there is always around my code. We also discussed other solutions but came to the conclusion that @agraul's solution with the dataclasses is the most elegant for now.